BoxOffice/Controller/BoxOfficeListViewController.swift

```diff
@@ -0,0 +1,145 @@
+//
+//  BoxOfficeListViewController.swift
+//  BoxOffice
+//
+//  Created by Rhode, Rilla on 2023/03/20.
+//
+
+
+import UIKit
+
+final class BoxOfficeListViewController: UIViewController {
+    private let server = NetworkManager()
+    private let urlMaker = URLMaker()
+    private var boxOffice: BoxOffice?
+    
+    private let collectionView: UICollectionView = {
+        let layout = UICollectionViewFlowLayout()
+        layout.minimumLineSpacing = 0
+        
+        layout.scrollDirection = .vertical
+        layout.sectionInset = UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 0)
+        
+        let collectionView = UICollectionView(frame: .zero, collectionViewLayout: layout)
+        collectionView.register(CustomCollectionViewCell.self,
+                                forCellWithReuseIdentifier: CustomCollectionViewCell.identifier)
+        
+        return collectionView
+    }()
+    
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        configureView()
+        configureRefreshControl()
+    }
+    
+    private func configureView() {
+        view.backgroundColor = .white
+        self.title = Date.configureYesterday(isFormatted: true)
+        LoadingIndicator.showLoading()
+        
+        collectionView.dataSource = self
+        collectionView.delegate = self
+        
+        fetchBoxOfficeData {
+            DispatchQueue.main.async {
+                LoadingIndicator.hideLoading()
+                self.collectionView.reloadData()
+            }
+        }
+        
+        view.addSubview(collectionView)
+        collectionView.setAutoLayout(equalTo: view.safeAreaLayoutGuide)
+    }
+    
+    private func fetchBoxOfficeData(completion: @escaping () -> Void) {
+        guard let url = urlMaker.makeBoxOfficeURL(date: Date.configureYesterday(isFormatted: false)) else { return }
+        server.startLoad(url: url) { result in
+            let decoder = DecodeManager()
+            
+            guard let verifiedFetchingResult = self.verifyFetchingResult(result: result) else { return }
+            let decodedFile = decoder.decodeJSON(data: verifiedFetchingResult, type: BoxOffice.self)
+            let verifiedDecodingResult = self.verifyDecodingResult(result: decodedFile)
+            
+            self.boxOffice = verifiedDecodingResult
+            completion()
+        }
+    }
+    
+    private func verifyFetchingResult(result: Result<Data, NetworkError>) -> Data? {
+        switch result {
+        case .success(let data):
+            return data
+        case .failure(_):
+            return nil
+        }
+    }
+    
+    private func verifyDecodingResult<T: Decodable>(result: Result<T, DecodeError>) -> T? {
+        switch result {
+        case .success(let data):
+            return data
+        case .failure(_):
+            return nil
+        }
+    }
```

Comment:
질문 주신 부분인데, 파라미터에서 타입을 명시하고 있으니 제네릭으로 묶을 수 있을 것 같습니다.

```suggestion
 private func verifyResult<T>(result: Result<T, Error> -> T? 
```

Reply:
지성이 조언해주신대로 묶어보려고 했으나 다음과 같은 오류가 납니다:

![](https://i.imgur.com/JGkHrZJ.png)

NetworkError를 any Error로 변환할 수 없다고 하네요.ㅠㅠㅠ

Reply:
Error도 제네릭으로 바꾸면 되지 않을까 싶은데 어떠신가요? <E: Error> 이런식으로요

Reply:
196523c9a313b4c53b9955f2db4451eb4e8b46a2

타입과 에러를 다 제너릭을 사용해 해결했습니다,,,! 생각보다 되게 간단한 작업이었는데 못하고 있었네요,,,